src/draw/sw/lv_draw_sw.c

```diff
@@ -113,7 +113,7 @@ static int32_t lv_draw_sw_dispatch(lv_draw_unit_t * draw_unit, lv_layer_t * laye
 
 #if LV_USE_OS
     /*Let the render thread work*/
-    lv_thread_sync_signal(&draw_sw_unit->sync);
+    if(draw_sw_unit->inited) lv_thread_sync_signal(&draw_sw_unit->sync);
```

Comment:
Not only the situation you mentioned, but also the thread scheduling strategies of various RTOSes are different, and there may be different performance behaviors even with the same priority.

Reply:
```
else
    return 0;
```
You can not return 1 if the condition is not true.

Reply:
Or `return -1` to ask LVGL to dispatch again.

Reply:
I was thinking maybe another unit is available to take it. -1 shall work too.

Reply:
Still returning 1 because we have configured `task_act` for draw unit, which means once the thread starts successfully, `execute_drawing` in `render_thread_cb` will be triggered without waiting for any condition."

Reply:
No, because how would you know next time if your unit is busy with other task in order to return 0?

Reply:
When `draw_act` is not consumed and then dispatched again, it will return 0.

Reply:
`if(draw_sw_unit->inited == false)`  the unit shouldn't take any tasks, as it's not initialized. So the `inited` flag should be checked very early.

Reply:
Please see Line 139. So we can dispatch it before draw unit thread's ready.

Reply:
Assuming that returning -1 directly maybe cause the main thread is in a busy loop. If returning 0, the current processing logic is the same as returning 1, but semantically it is not consistent with the case where the task is first taken and 1 is returned.

Reply:
I was writing a long comment about why I think I'm right, and in the meantime I realized that you are right. :slightly_smiling_face: 

I agree to return 1. 

Just to be sure we are on the same page this is what happens:
1. The render thread is not initialized yet, but `u->task_act` is set in the dispatcher.
2. the dispatcher returns 1 to indicate it taken a task
3. When the render thread will run, it will initialize the sync and due to the code below it will process it immediately:
```c
        while(u->task_act == NULL) {
            lv_thread_sync_wait(&u->sync);
        }
```

Reply:
@bjsylvia how does it look your dispatcher code? Can you please post it here?